api/src/user/user.handlers.ts

```diff
@@ -140,3 +115,82 @@ export const getUserWorksheet = async (
     data: worksheetCoursesToWorksheets(worksheets)[netId] ?? {},
   });
 };
+
+export const updateBookmark = async (
+  req: express.Request,
+  res: express.Response,
+): Promise<void> => {
+  winston.info('Updating course bookmark');
+
+  const { netId } = req.user!;
+
+  const bodyParseRes = UpdateBookmarkReqBodySchema.safeParse(req.body);
+  if (!bodyParseRes.success) {
+    winston.info(bodyParseRes.error.issues);
+    res.status(400).json({ error: 'INVALID_REQUEST' });
+    return;
+  }
+
+  bodyParseRes.data.forEach(async (update) => {
+    const { action, season, crn, worksheetNumber, color, hidden } = update;
+
+    const [existing] = await db
+      .selectDistinctOn([
+        worksheetCourses.netId,
+        worksheetCourses.crn,
+        worksheetCourses.season,
+        worksheetCourses.worksheetNumber,
+      ])
+      .from(worksheetCourses)
+      .where(
+        and(
+          eq(worksheetCourses.netId, netId),
+          eq(worksheetCourses.crn, crn),
+          eq(worksheetCourses.season, season),
+          eq(worksheetCourses.worksheetNumber, worksheetNumber),
+        ),
+      );
+
+    if (!existing) {
+      res.status(400).json({ error: 'NOT_BOOKMARKED' });
```

Comment:
My current plan is to still commit the rest of the changes and report partial failure, because there's no dependency between these updates, and it may be better to do _something_ rather than _nothing_.